In [1]:
import pandas as pd
import numpy as np
import collections
import ngram
import matplotlib.pyplot as plt
import seaborn as sns
sns.set_style('whitegrid')

In [2]:
# Load data
f = open('./data/j101-d_11_1499.txt', encoding='shift-jis')
line = f.read()
f.close()
print(line[0:300])

エレキベースのスラップ奏法を対象としたスペクトル特徴と演奏制約を利用した音高・奏法推定あらまし本論文では，エレキベースの奏法の違いによる音高の推定精度の低下を改善するために，特殊な奏法の一つであるスラップ奏法を対象にスペクトル特徴と演奏制約を導入し，ネットワークの最短経路を求めることで音高と奏法を推定する手法を提案する．キーワードエレキベース，調波打楽器音分離，スラップ奏法，非負値行列因子分解１．まえがき近年，計算機を用いて音楽信号から自動で楽譜を作成する自動採譜に関する研究が盛んに行われている［１］．自動採譜システムの実現のためには，楽曲の調やコードと関連が強いベースラインの推定が重要となる


## Q1

In [3]:
# Q1
print('Total number is', len(line))
print('Total kind of character is', len(collections.Counter(line)))

Total number is 5960
Total kind of character is 450


上記プログラムより、延べ文字数5960文字、文字種は450文字である。

## Q2

In [4]:
# Q2
print('H0 =', np.log2(len(collections.Counter(line))))

H0 = 8.813781191217037


In [5]:
cnt = collections.Counter(line)
freq = np.array(list(cnt.values()))
cnt.values() # words frequency

dict_values([20, 35, 17, 21, 71, 81, 255, 28, 65, 35, 129, 79, 153, 22, 2, 132, 111, 80, 17, 48, 80, 29, 22, 14, 64, 22, 22, 4, 27, 108, 50, 3, 30, 41, 29, 29, 22, 5, 4, 6, 87, 48, 181, 20, 57, 132, 31, 174, 7, 34, 9, 5, 1, 1, 96, 22, 1, 51, 6, 14, 12, 13, 13, 20, 16, 5, 4, 4, 8, 43, 17, 10, 10, 116, 16, 11, 22, 6, 24, 10, 20, 18, 2, 2, 12, 6, 3, 5, 3, 9, 52, 9, 79, 23, 4, 1, 10, 11, 1, 70, 6, 13, 20, 4, 19, 6, 4, 14, 4, 13, 1, 1, 1, 2, 14, 36, 12, 12, 5, 3, 30, 12, 2, 8, 11, 18, 7, 6, 16, 13, 6, 1, 45, 2, 2, 5, 7, 29, 1, 1, 24, 14, 3, 5, 12, 7, 8, 1, 1, 9, 7, 8, 16, 5, 5, 3, 17, 34, 8, 22, 12, 6, 7, 12, 11, 3, 3, 11, 25, 4, 4, 3, 5, 11, 11, 30, 12, 15, 17, 4, 8, 52, 61, 10, 3, 8, 3, 14, 26, 12, 9, 6, 10, 1, 4, 4, 3, 9, 9, 19, 52, 13, 2, 6, 18, 9, 11, 15, 7, 1, 15, 2, 2, 3, 7, 2, 9, 12, 19, 12, 3, 4, 1, 1, 1, 1, 23, 5, 6, 2, 4, 30, 1, 2, 22, 12, 22, 28, 1, 1, 5, 1, 1, 5, 4, 2, 24, 6, 4, 10, 3, 5, 2, 4, 8, 2, 2, 1, 1, 1, 1, 5, 2, 5, 2, 7, 16, 5, 5, 2, 2, 5, 1, 1, 2, 1, 5, 5, 5, 2, 18, 3

In [6]:
h1 = 0
for i in range(len(freq)):
    h1 += (freq[i]/len(line)) * np.log2(freq[i]/len(line))
h1 *=  -1
print('H1 =', h1)
print('Information gain: H0-H1 =', np.log2(len(collections.Counter(line)))-h1)

H1 = 7.459371410966394
Information gain: H0-H1 = 1.3544097802506432


(a) 全ての文字種が等確率で出現すると仮定した場合、文字種が450であることから、エントロピーは以下の式で求めることができる。

$$
H_0 = \log_2 450 \\
H_0 = 8.813781191217037
$$

(b) 文字の出現頻度を考慮した際のエントロピーは以下の通りである。
$$
H_1 = -\frac{20}{5960} -\log_2 \frac{20}{5960} -\frac{35}{5960} -\log_2 \frac{35}{5960} ... =7.459371410966394
$$

(c) 以上の結果より、Information gain (IG)は
$$
IG = H_0 - H_1 = 1.3544097802506432
$$
となり、約1.35[bit]である。

情報エントロピーは、事前に知っていることが多いほど情報量は少なくなる。言い換えれば、不確実性が少なければ情報量も少なくなる。そのため、出現頻度を考慮したH<sub>1</sub>が、H<sub>0</sub>よりも不確実性が低いと言える。

## Q3

In [7]:
# Q3
# (value of kind of characters * 2byte)^3 
print((len(collections.Counter(line))*2)**3, 'byte')

729000000 byte


In [8]:
# 2l+4l+l=7l 
# 7l * text size memmory must be reserbed
print(7 * len(collections.Counter(line)), 'byte')

3150 byte


(a) 全文字種類は450文字であり、これらは全て全角2 byteからなる。これらの文字に対してtri-gramを適用することを考慮した場合必要となる容量は以下のとおりである。
$$
(450 \times 2)^3 = 729000000 [bytes]
$$

(b) Nagaoらの手法では、text string(2l byte), pointer table(4l bytes), table of coincidence number of characters(1l byte)の計7l bytesが必要となる。このことから、(a)と同様に、全文字種類についてtri-gramを適用すると
$$
7 \times 450 = 3150 [bytes]
$$
となり、3150 bytesのメモリを確保する必要がある。

(c) 以上の結果から、Nagaoらの手法を用いることで、確保するメモリ容量を大幅に削減できることが分かる。通常のN-gramではNの数が大きくなればなるほど、容量は大きくなるが、Nagaoらの手法では、ポインターを用いて参照を行うため、大規模なN-gramであっても容量の削減が期待できる。

## Q4

In [9]:
# create n-gram
index = ngram.NGram(N=6)
ngram = []

for term in index.ngrams(index.pad(line)):
    ngram.append(term)
ngram[0:10]

['$$$$$エ',
 '$$$$エレ',
 '$$$エレキ',
 '$$エレキベ',
 '$エレキベー',
 'エレキベース',
 'レキベースの',
 'キベースのス',
 'ベースのスラ',
 'ースのスラッ']

In [10]:
dictionary = dict()
lcw = []

for s1, s2, s3, s4, s5, s6 in ngram:
    dictionary[(s1, s2, s3, s4, s5, s6)] = dictionary.get((s1, s2, s3, s4, s5, s6),0)+1
    
for i in range(len(dictionary)):
    if list(dictionary.values())[i] > 15:
        lcw.append(list(dictionary)[i])
        print(list(dictionary)[i], list(dictionary.values())[i])        

('エ', 'レ', 'キ', 'ベ', 'ー', 'ス') 16
('ス', 'ラ', 'ッ', 'プ', '奏', '法') 17


プログラムより、閾値15かつ、6文字以上を満たす the longest compound wordとして、以下の2つが抽出された。

+ エレキベース (16)
+ スラップ奏法 (17)

## Q5

In [11]:
# split word to w1 and w2
w1, w2 = [], []

for i in range(len(lcw)):
    word = ''.join(lcw[i])
    
    for j in range(len(word)-1):
        #print(word[:j+1], word[j+1:])
        w1.append(word[:j+1])
        w2.append(word[j+1:])
print(w1)
print(w2)

['エ', 'エレ', 'エレキ', 'エレキベ', 'エレキベー', 'ス', 'スラ', 'スラッ', 'スラップ', 'スラップ奏']
['レキベース', 'キベース', 'ベース', 'ース', 'ス', 'ラップ奏法', 'ップ奏法', 'プ奏法', '奏法', '法']


In [12]:
# count freq of w1 and w2
w1_freq = []
w2_freq = []
w_sum = []

for i in range(len(w1)):
    w1_freq.append(line.count(w1[i]))
    w2_freq.append(line.count(w2[i]))
    w_sum.append(line.count(w1[i]) + line.count(w2[i]))
    
df = pd.DataFrame([w1, w1_freq, w2, w2_freq, w_sum]).T
df.columns = ['w1', 'w1 frequency', 'w2', 'w2 frequency', 'sum']

以下が分割したw1, w2とそれぞれの出現頻度をまとめた表である。

In [13]:
df

w1 w1 frequency     w2 w2 frequency sum
0      エ           20  レキベース           16  36
1     エレ           16   キベース           16  32
2    エレキ           16    ベース           19  35
3   エレキベ           16     ース           24  40
4  エレキベー           16      ス           81  97
5      ス           81  ラップ奏法           17  98
6     スラ           18   ップ奏法           17  35
7    スラッ           17    プ奏法           17  34
8   スラップ           17     奏法           65  82
9  スラップ奏           17      法           79  96

　頻度の合計が最も大きくなった組み合わせは ['エレキベー'(16), 'ス'(81)], ['ス'(81), 'ラップ奏法'(17)] であった。適当な分割として、妥当に思えるw<sub>1</sub>とw<sub>2</sub>は、['エレキ'(16), 'ベース'(19)], ['スラップ'(17), '奏法'(65)] である。  
　文字数が小さいほど、文中で登場する回数は多くなる可能性が高く、今回このような結果を得たのもこれが理由であると考えられる。実際に、['エ', 'レキベース'], ['エレキベー', 'ス'], ['ス', 'ラップ奏法'], ['スラップ奏', '法']のように、1文字のみになる分割を行った際の頻度が高くなっている。